### PR TITLE
Throttle ProxySQL Connection Attempts to Mitigate Connection Attempt Spikes

### PR DESCRIPTION
**Pull request by sureshc**
Throttle ProxySQL Connection Attempt spikes that appear to be causing availability events.

https://codedotorg.atlassian.net/browse/INF-1429

Since cutting over to Aurora 3 / MySQL 8, the [highest Connection Attempt spike to the writer database instance](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'writer_connections*20*2f*20web_server_instances~label~'Writer*20Connections*20Per*20Web*20Server~id~'writer_connections_per_instance~region~'us-east-1~period~3600~visible~false))~(~(expression~'writer_connection_attempts*20*2f*20web_server_instances~label~'Writer*20Connection*20Attempts*20Per*20Web*20Server~id~'writer_connection_attempts_per_instance~region~'us-east-1~period~3600~yAxis~'right~color~'*23aec7e8))~(~(expression~'reader_connections*20*2f*20web_server_instances~label~'Reader*20Connections*20Per*20Web*20Server~id~'reader_connections_per_instance~region~'us-east-1~period~3600~color~'*23ff7f0e~visible~false))~(~(expression~'reader_connection_attempts*20*2f*20web_server_instances~label~'Reader*20Connection*20Attempts*20Per*20Web*20Server~id~'reader_connection_attempts_per_instance~region~'us-east-1~period~3600~yAxis~'right~color~'*23ffbb78~visible~false))~(~'AWS*2fRDS~'DatabaseConnections~'Role~'WRITER~'DBClusterIdentifier~'autoscale-prod-cluster~(region~'us-east-1~id~'writer_connections~visible~false))~(~'AWS*2fAutoScaling~'GroupInServiceInstances~'AutoScalingGroupName~'Frontends-autoscale-prod~(id~'web_server_instances~label~'Autoscale*20InService~region~'us-east-1~visible~false))~(~'AWS*2fRDS~'DatabaseConnections~'Role~'READER~'DBClusterIdentifier~'autoscale-prod-cluster~(region~'us-east-1~id~'reader_connections~visible~false))~(~'.~'ConnectionAttempts~'.~'WRITER~'.~'.~(id~'writer_connection_attempts~yAxis~'right~color~'*23aec7e8~region~'us-east-1~visible~false))~(~'...~'READER~'.~'.~(id~'reader_connection_attempts~yAxis~'right~color~'*23ffbb78~region~'us-east-1~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Maximum~period~3600~title~'DB*20Connections*20Per*20Web*20Server~yAxis~(left~(min~0))~start~'-PT2160H~end~'P0D)) has been ~500 / hour / web application server EC2 Instance (less than one per second per instance), but if only ProxySQL on one or two EC2 Instances is responsible for a Connection Attempt spike, start with a conservative throttle at 100 per web application server instance.

image

## Testing story

Successfully provisioned a "full stack" adhoc (auto scaling web application servers with ProxySQL installed):
https://adhoc-throttle-proxysql2-studio.cdn-code.org/users/sign_in

## Deployment strategy

If this initial setting negatively impacts performance or availability, we can change the setting at runtime in ProxySQL all production web application servers using the `production-daemon` `bin/fanout` script. For example, this command sets the throttle to 200 and outputs the updated setting:

`bin/fanout "~/adhoc/bin/admin-proxysql 'SET mysql-throttle_connections_per_sec_to_hostgroup=200;SELECT @@mysql-throttle_connections_per_sec_to_hostgroup;'"`

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb [hunk 1]
```diff
@@ -82,6 +82,13 @@ mysql_variables=
   # https://github.com/sysown/proxysql/wiki/Global-variables#mysql-ping_timeout_server
   # [Milliseconds, default 200]
   ping_timeout_server=1000
+
+  # Prevent Connection Attempt spikes - a negative feedback cycle where a short period of high DB Load due to high usage
+  # or inefficient queries increases query latency, which increases the likelihood that ProxySQL needs to open new
+  # connections to execute new queries, which in turn further degrades query latency due to CPU load to service
+  # new connection attempts.
+  # https://proxysql.com/documentation/global-variables/mysql-variables/#mysql-throttle_connections_per_sec_to_hostgroup
+  throttle_connections_per_sec_to_hostgroup = <%= throttle_connections_per_sec_to_hostgroup  %>
```
Comment (Hamms) [inline cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb]: Does this not need to be
```suggestion
  throttle_connections_per_sec_to_hostgroup = <%= @throttle_connections_per_sec_to_hostgroup  %>
```

?

Comment (Hamms) [reply cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb]: Or, even better:

```suggestion
  throttle_connections_per_sec_to_hostgroup = <%= @throttle_connections_per_sec_to_hostgroup %>
```

**Review by Hamms — COMMENTED** [review]
I can't say I have any idea what the potential performance implications of this are, but the changes look good to me from a clarity/readability perspective and I only have one tiny question about syntax.

The `bin/fanout` response plan in the event that this does have negative implications sounds reasonable, but just to confirm: that would change the setting for all existing servers, but any new ones we autoscale up to will default back to 100, correct?

**Comment by sureshc** [comment]
> The `bin/fanout` response plan in the event that this does have negative implications sounds reasonable, but just to confirm: that would change the setting for all existing servers, but any new ones we autoscale up to will default back to 100, correct?

Correct - using fanout to change ProxySQL runtime configuration just buys us time until we override the Chef Attribute and trigger a build with an empty commit to provision a new AMI to permanently change the setting for all app servers.

File: cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb (hunk 1 above)
Comment (sureshc) [reply cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb]: oops. good catch! Fortunately the existing syntax worked, but how would you feel about just referencing the attribute directly? We do this at least one other place in this template. I’m firing up a new adhoc to test that out.

https://github.com/code-dot-org/code-dot-org/blob/a145c11a95cf7b1470f10c6082f5efd6eb13b386/cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb#L91

Comment (sureshc) [reply cookbooks/cdo-mysql/templates/default/proxysql.cnf.erb]: I successfully provisioned a "full stack" adhoc with the above change that would just reference the attribute directly in the template that renders the ProxySQL configuration file:

```
mysql> SELECT @@mysql-throttle_connections_per_sec_to_hostgroup;
+---------------------------------------------------+
| @@mysql-throttle_connections_per_sec_to_hostgroup |
+---------------------------------------------------+
| 100                                               |
+---------------------------------------------------+
1 row in set (0.00 sec)
```

**Review by Hamms — APPROVED** [review]
Direct reference makes sense to me, and it's definitely a cleaner diff! LGTM, thanks

**Pull request merged**